### PR TITLE
Fix signature for Enumerable#detect

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

Fixes https://github.com/sorbet/sorbet/issues/3240.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HwWb2R8VRjYlfz

File: rbi/core/enumerable.rbi [hunk 1]
```diff
@@ -374,17 +374,19 @@ module Enumerable
...
   sig do
-    params(
-        ifnone: Proc,
-    )
-    .returns(T::Enumerator[Elem])
+    type_parameters(:U)
+      .params(
+        ifnone: T.proc.returns(T.type_parameter(:U)),
+      )
+      .returns(T::Enumerator[T.any(T.type_parameter(:U), Elem)])
```
Comment (elliottt) [inline rbi/core/enumerable.rbi:389]: This raises some issues on the stripe codebase, as leaving off the `ifnone` block parameter means that the result will just be nilable. I think you probably just need a third overload that doesn't include it. Could you add a test for this case:

```
T.reveal_type([1,2].detect {|x| false}) # error: Revealed type `T.nilable(Integer)`
```

Comment (Morriar) [reply rbi/core/enumerable.rbi:389]: Good catch. I fixed this and added a bunch of tests in `test/testdata/rbi/enumerable.rb` 👍

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HzFgqDEjlSYK9d

**Review by elliottt — APPROVED** [review]
Thanks!

**Pull request merged**